<!DOCTYPE html>
<html>
  <head>
    <title>P-values in quality control of high-throughput experiments</title>
    <meta charset="utf-8">
    <meta name="date" content="2017-06-08" />
    <link rel="stylesheet" href="bc.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# P-values in quality control of high-throughput experiments
### <p>Taavi Päll<br />
Institute of Chemistry and Biotechnology<br />
Tallinn University of Technology<br />
<a href="mailto:taavi.pall@ttu.ee">taavi.pall@ttu.ee</a></p>
### 2017-06-08

---





class: inverse, center, middle

# Let's test how SRP work in real life 

---

### The GEO database

- GEO accepts next generation sequence data that examine quantitative gene expression, gene regulation, epigenomics or other aspects of functional genomics using methods such as RNA-seq, miRNA-seq, ChIP-seq, RIP-seq, HiC-seq, methyl-seq, etc

&lt;img src="index_files/figure-html/geo-dynamics-1.svg" style="display: block; margin: auto;" /&gt;

---

### Query

&gt; 'expression profiling by high throughput sequencing[DataSet Type] AND ("Homo sapiens"[Organism] OR "Mus musculus"[Organism])'

- query retrieved 8670 GEO Series (Apr-24, 2017)

![query](figs/GEO-query.png)


---

### Supplemental files

```r
suppfilenames &lt;- ds %&gt;%
*   filter(PDAT&gt;"2015/12/24", PDAT&lt;"2017/03/31",
         !inherits(SuppFileNames, "try-error"))

*   out &lt;- c("tar","gtf","bed","bigbed","bedgraph","bw","wig","hic","gct(x)?","tdf",
*         "gff","pdf","png","zip","sif","narrowpeak","gff3","fa")

suppfiles_of_interest &lt;- suppfilenames %&gt;%
  unnest(SuppFileNames) %&gt;%
*  filter(!str_detect(tolower(SuppFileNames),
*  "filelist|error|readme|annotation|raw.tar$|[:punct:]hic|hdf5$"),
         !str_detect(tolower(SuppFileNames), paste0(out, "(\\.gz)?$", collapse = "|")))
```

- GEO Series published between Dec-24 2015 and Mar-31 2017
- Keep only tables
- Try to exclude single cell seq experiments (some will sneak through, as people are really inventive in file (mis)labeling), Chr-IP peak files, pdf, png, readmes etc.

---

### GEO Series with (potentially interesting) supplemental files

&lt;img src="index_files/figure-html/suppfilesofinterest-1.svg" style="display: block; margin: auto;" /&gt;

---

### Number of samples and features in high-throughput sequencing experiments

.pull-left[

&lt;table&gt;
 &lt;thead&gt;
  &lt;tr&gt;
   &lt;th style="text-align:left;"&gt; stat &lt;/th&gt;
   &lt;th style="text-align:right;"&gt; features &lt;/th&gt;
   &lt;th style="text-align:right;"&gt; samples &lt;/th&gt;
  &lt;/tr&gt;
 &lt;/thead&gt;
&lt;tbody&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; mean &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 65051 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 36 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; median &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 25546 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 9 &lt;/td&gt;
  &lt;/tr&gt;
&lt;/tbody&gt;
&lt;/table&gt;

]

.pull-right[
&lt;img src="index_files/figure-html/samples-1.svg" style="display: block; margin: auto;" /&gt;
]

.footnone[
Summary table, threshold of 8000 rows/features was used to calculate mean and median. Features are transcripts, genomic locuses, etc. GEO series with ~ 1000 samples are single-cell sequencing experiments.
]

---

### We were looking for supplemental files including tables with:

1. Unadjusted raw P values
2. Raw sequence read counts (not CPM, RPKM)

&lt;img src="index_files/figure-html/pvalues-1.svg" style="display: block; margin: auto;" /&gt;

.footnone[
Threshold of 8000 was used to identify full expressome P values sets.
]

---

### GEO Series with raw sequence read counts

![esets](figs/esets.png)

96 unique GEO Series

---

![recount2](figs/recount2-screen.png)

---

### Fit models


```r
# Identify vars for model fitting
esets %&lt;&gt;% mutate(groups = map(result, ~get_model_eset(pData(.x))))
library(edgeR); library(tidyverse); library(magrittr)
esets %&lt;&gt;% filter(map_lgl(result, ~nrow(exprs(.x))&gt;8000)) %&gt;%
  mutate(dge = pmap(list(result, groups, countfiles),
    function(x,y,z) {message(z); DGEList(counts=exprs(x), samples = y, genes = fData(x))})) %&gt;%
  filter(!map_lgl(groups, is.null))
# Make design formulas and model.matrix
make_formula &lt;- . %&gt;% colnames %&gt;% paste(collapse="+") %&gt;% paste("~",.) %&gt;% formula()
esets %&lt;&gt;% mutate(design = map(groups, make_formula))
esets %&lt;&gt;% mutate(design = map2(design, groups, ~model.matrix(object=.x, data=.y)))
# Filter genes with low counts: 3 or more samples with cpm of 10/L. L, # millions of counts in the smallest library
esets %&lt;&gt;% mutate(keep = map(dge, ~rowSums(cpm(.x)&gt;10/min(.x$samples$lib.size)/1e6) &gt;= 3),
                  dge_filt = map2(dge, keep, ~.x[.y,]))
# Recalculate library sizes and calculate normalisation factors
esets %&lt;&gt;% mutate(dge_filt = map(dge_filt, ~{.x$samples$lib.size &lt;- colSums(.x$counts); .x}))
esets %&lt;&gt;% mutate(dge_filt = map2(countfiles, dge_filt, ~{message(.x); calcNormFactors(.y)}))
esets %&lt;&gt;% mutate(dge_filt = pmap(list(dge_filt, design, countfiles),
                                  function(x,y,z) {message(z); try(estimateDisp(x,y))}))
# Fit models
esets %&lt;&gt;% filter(!map_lgl(dge_filt, ~inherits(.x, "try-error"))) %&gt;%
  mutate(fit = pmap(list(dge_filt, design, countfiles), function(x,y,z) {message(z); try(glmFit(x,y))})) %&gt;%
  filter(!map_lgl(fit, ~inherits(.x, "try-error"))) %&gt;%
  mutate(results = map(fit, ~glmLRT(.x, coef=2:ncol(.x$design))),
         toptags = map(results, topTags, n=Inf))
```

---

### Interpreting P value histograms

.pull-left[
- Make a histogram of your P values
- Look closely at your P value histograms as a first step in quality control
&gt; Do this before you perform multiple hypothesis test correction, false discovery rate control, or any other means of interpreting your many P values&lt;sup&gt;1&lt;/sup&gt;
]

.pull-right[
&lt;img alt="good and bad P value histograms"  src="http://varianceexplained.org/figs/2014-12-15-interpreting-pvalue-histogram/plot_melted-1.png" height="420"&gt;
]

.footnote[
&lt;sup&gt;1&lt;/sup&gt;How to interpret a P value histogram: [http://varianceexplained.org/statistics/interpreting-pvalue-histogram](http://varianceexplained.org/statistics/interpreting-pvalue-histogram/)
]

???

- P value histogram lets you get an immediate sense of how your test behaved across all your hypotheses, and immediately diagnose some potential problems

---

### Be careful with pathological cases

.pull-left[
Big bump in the middle? Bunch of random peaks? 
![weird](http://varianceexplained.org/figs/2014-12-15-interpreting-pvalue-histogram/weird_pvalues-1.png)
]

.pull-right[
&lt;img src="https://1.bp.blogspot.com/-ff59ebmURDE/UYyLvQUURGI/AAAAAAAABXc/z7OWx8Mus5Q/s1600/don%27twww-scarfolk-blogspot-com.jpg" height="420"&gt;
]

.footnote[
Images: left, varianceexplained.org; right, Scafolk Council.
]

---

class: inverse, center, middle

# Results

---

### Function to calculate srp


```r
devtools::install_github("tpall/SRP")
library(SRP)
pvals &lt;- runif(1000)
out &lt;- try(srp(pvals))
out
```

```
##   SRP       pi0 fp rs       ud
## 1   0 0.8794309  0  0 120.5691
```


---
### P value histograms, estimated pi0, and stable retrospective power

![SRP table](figs/SRP-table.png)

[Click for full table](pvalue-histograms.html)

.footnote[
P values are from likelihood ratio tests where full model formula was compared against intercept-only formula.
]

---

### SRP versus proportion of true nulls

&lt;img src="index_files/figure-html/srp-scatterplot-1.svg" style="display: block; margin: auto;" /&gt;


---

### SRP versus library size

&lt;img src="index_files/figure-html/srp-libsize-1.svg" style="display: block; margin: auto;" /&gt;


---

### Acknowledgements

- Ülo Maiväli and Tanel Tenson, _Institute of Technology, University of Tartu_
- Read more about GEO RNA-seq project in [blog.credibleinterval.ee](https://blog.credibleinterval.ee)


.footnote[
email: taavi.pall@ttu.ee
]
    </textarea>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightLines": true
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {window.dispatchEvent(new Event('resize'));});</script>

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  tex2jax: {
    skipTags: ['script', 'noscript', 'style', 'textarea', 'pre']
  }
});
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
